CMakeLists.txt

```diff
@@ -355,6 +358,11 @@ set(BROKER_SRC
   src/telemetry/metric_registry_impl.cc
   src/time.cc
   src/topic.cc
+  src/variant.cc
+  src/variant_data.cc
+  src/variant_list.cc
+  src/variant_set.cc
+  src/variant_table.cc
```

Comment:
I would like to keep the API close to `broker::data`, in API but also in terms of run-time properties`*`. Having just lists internally bakes in assumptions that we have about client code that in the end might not be true and could severely backfire. Basically, I would like to first try a design that is as close as possible to `broker::data`. We can still pick up your idea later if we need to fine-tune performance further. The good thing is that we have distinct wrapper classes now that hide the underlying data type. This gives us quite a bit of leeway to change the internal representation later on.

`*` a `set` has `O(log n)` lookup complexity, same as a `map`. What usually makes these containers slow are memory allocations and pointer chasing. However, this is no longer true in a design that allocates everything into a big memory block. I think this renders the discussion with `std::map` vs `std::unordered_map` obsolete. IIRC, we had the `Rb_tree` from this flame graph: https://user-images.githubusercontent.com/2100509/188454005-3456b0e3-0c87-4ad5-b6c6-dab3bb00eb0c.svg. The biggest chunk here is heap allocations. This will be much cheaper with the custom allocation strategy we have in place for `broker::variant`. In fact, we probably want to avoid `unoredered_map` here because it will frequently re-allocate smaller chunks that would fragment our monotonic buffer resource and create gaps. Same reason we avoid `std::vector` here and actually use a `std::list` (which is usually much slower than a `vector`).

Reply:
Cool, thanks for the info and pointing out the differences in memory allocation.